### PR TITLE
Add Peer Limit to Swarm

### PR DESCRIPTION
**Pull request by nisdas**
This PR adds a peer limit to the Swarm. The purpose of this is to limit new connections/streams being opened with peers once the limit is reached. For my usecase I needed to enforce a peer limit so that my libp2p host would not be connected to more peers than specified. However I could not find a way to configure and enforce this with the current swarm object. If there is a better way to achieve this I would be all ears.

For your reference, this is another branch that adds in a Peer Limit Option, in `go-libp2p` 
https://github.com/libp2p/go-libp2p/compare/master...nisdas:peerLimiter

### CODE REVIEW

**Comment by aarshkshah1992** [comment]
@nisdas 

I think [connection manager](https://github.com/libp2p/go-libp2p-connmgr) is what you are looking for. Here's a code sample I used in one of my projects

```
connManager := connmgr.NewConnManager(nMinConnections, nMaxConnections)
opts := []libp2p.Option{
  libp2p.ConnectionManager(connManager),
}
host, err := libp2p.New(ctx, opts...)
```

Let me know if you have any questions.

**Comment by nisdas** [comment]
Unfortunately, connection manager does not satisfy my usecase, its a reactive solution to what I need. It still opens connections with peers and only trims them at every interval. This leads to a lot of peer churn which destabilizes the network, I am trying to avoid connecting to any new peers completely and adding them to my peerstore.

**Comment by prestonvanloon** [comment]
If it’s helpful, we also filed a question on the forum a few days ago. 

https://discuss.libp2p.io/t/what-is-the-proper-way-to-limit-connections/357

File: swarm.go [hunk 1]
```diff
@@ -178,6 +182,10 @@ func (s *Swarm) Process() goprocess.Process {
 }
 
 func (s *Swarm) addConn(tc transport.CapableConn, dir network.Direction) (*Conn, error) {
+	numOfPeers := len(s.conns.m)
+	if numOfPeers >= int(s.peerLimit) && s.Connectedness(tc.RemotePeer()) != network.Connected && s.peerLimit != 0 {
+		return nil, errors.New("number of peers over the peer limit, so rejecting connection")
+	}
```
Comment (vyzo) [inline swarm.go]: Two things wrong with this:
- you need a default of 0 to mean no peer limit, so that thing work as they currently do (this should be the default)
- you are leaking the connection here, you need to close it if you are rejecting.

Comment (nisdas) [reply swarm.go]: Hey the if the peer limit is set to zero it ignores this conditional. 
```go 
&& s.peerLimit != 0
```
 I added a var to make this more clear. 
I also now close the connection in the event of an error. 

If I want to set a fixed peer limit , after which no peer will be able to connect with us what would be the best way to go about this ?

Comment (vyzo) [reply swarm.go]: Put the peer limit not being 0 first please, this was hard to parse.
Also, it's the default, so it will be faster to short-circuit.

In addition, you should log something about this connection closure.

File: swarm.go [hunk 2]
```diff
@@ -181,6 +188,12 @@ func (s *Swarm) Process() goprocess.Process {
 }
 
 func (s *Swarm) addConn(tc transport.CapableConn, dir network.Direction) (*Conn, error) {
+	numOfPeers := len(s.conns.m)
+	nonZeroLimit := s.peerLimit > 0
+	if nonZeroLimit && numOfPeers >= int(s.peerLimit) && s.Connectedness(tc.RemotePeer()) != network.Connected {
+		tc.Close()
```
Comment (vyzo) [inline swarm.go]: Please log this.
Also, what's the purpose of the Connectedness check?

Comment (nisdas) [reply swarm.go]: Added in a log, the purpose of the Connectedness check is to continue accepting connections from already connected peers. We only reject connections from peers to which we have no active connections with.

Comment (vyzo) [reply swarm.go]: but you are not counting peers, you are counting connections.

Comment (nisdas) [reply swarm.go]: unless I am misunderstanding the code 
```go
len(s.conns.m)
```
wouldn't this represent all the peers with active connections currently ?

The type for `s.conns.m` is  
`map[peer.ID][]*Conn`

Comment (vyzo) [reply swarm.go]: oh that's true; sorry, my bad.

File: swarm.go [hunk 3]
```diff
@@ -181,6 +188,17 @@ func (s *Swarm) Process() goprocess.Process {
...
 
 func (s *Swarm) addConn(tc transport.CapableConn, dir network.Direction) (*Conn, error) {
+	p := tc.RemotePeer()
+	numOfPeers := len(s.conns.m)
+	nonZeroLimit := s.peerLimit > 0
+
+	// Check if the connection would exceed our specified peer limit.
+	if nonZeroLimit && numOfPeers >= int(s.peerLimit) && s.Connectedness(p) != network.Connected {
+		tc.Close()
+		log.Debugf("rejecting connection with peer [%s]", p)
```
Comment (vyzo) [inline swarm.go]: nit: "from peer", no need to put the peer ID in brackets.

File: swarm.go [hunk 4]
```diff
@@ -40,6 +40,9 @@ var ErrSwarmClosed = errors.New("swarm closed")
 // transport is misbehaving.
 var ErrAddrFiltered = errors.New("address filtered")
 
+// ErrPeerLimitExceeded is returned when we exceed the specified peer limit.
+var ErrPeerLimitExceeded = errors.New("number of peers over the peer limit, so rejecting connection")
```
Comment (vyzo) [inline swarm.go]: nit: the "so" is unnecessary, there is no conversation happening.

File: swarm.go [hunk 5]
```diff
@@ -97,12 +103,13 @@ type Swarm struct {
 }
 
 // NewSwarm constructs a Swarm
-func NewSwarm(ctx context.Context, local peer.ID, peers peerstore.Peerstore, bwc metrics.Reporter) *Swarm {
+func NewSwarm(ctx context.Context, local peer.ID, peers peerstore.Peerstore, bwc metrics.Reporter, peerLimit int) *Swarm {
```
Comment (vyzo) [inline swarm.go]: This is probably best done with functional options, so that we don't break existing code.

Comment (nisdas) [reply swarm.go]: Any idea on how to make that happen ? I couldn't find a swarm 'config' that I could use instead, so I had to add it as an additional argument.

Comment (vyzo) [reply swarm.go]: You can make the options take the swarm pointer itself.
So:
```
type Option func(s *Swarm) error
func NewSwarm(stuff, opt Option...)
```

**Comment by nisdas** [comment]
@vyzo addressed your comments and also added in a test case, let me know what you think

File: swarm_options.go [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package swarm
+
+// Option is a Swarm Option that can be given to a Swarm Constructor(`NewSwarm`).
+type Option func(s *Swarm)
```
Comment (vyzo) [inline swarm_options.go]: let's also return an error from the option.

File: swarm_options.go [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+func (s *Swarm) ApplyOptions(opts ...Option) {
+	for _, opt := range opts {
+		if opt == nil {
```
Comment (vyzo) [inline swarm_options.go]: that seems complete unnecessary.

File: swarm_options.go (hunk 6 above)
Comment (nisdas) [reply swarm_options.go]: the reason I didn't return an error was because if we did, then the
swarm constructor also has to return an error
```go
 NewSwarm(ctx context.Context, local peer.ID, peers peerstore.Peerstore, bwc metrics.Reporter, opts ...Option) (*Swarm,error) {
```
This would break existing code, if you still want me to change, I will 
add it in

Comment (vyzo) [reply swarm_options.go]: Ah yes, good point; let's not.

Comment (vyzo) [reply swarm_options.go]: or we can panic on error, but that's something we can do from the option itself I guess.

File: swarm_test.go [hunk 8]
```diff
@@ -61,7 +61,7 @@ func makeDialOnlySwarm(ctx context.Context, t *testing.T) *Swarm {
 	return swarm
 }
 
-func makeSwarms(ctx context.Context, t *testing.T, num int, opts ...Option) []*Swarm {
+func makeSwarms(ctx context.Context, t *testing.T, num int, opts ...TestOption) []*Swarm {
```
Comment (vyzo) [inline swarm_test.go]: what's this `TestOption`?

Comment (vyzo) [reply swarm_test.go]: ah found it, nm.

**Comment by vyzo** [comment]
travis test failure because of a data race.

**Comment by nisdas** [comment]
@vyzo fixed the data race, sounds good I will be waiting for @Stebalien review

**Comment by nisdas** [comment]
@raulk 
Hey thanks for the detailed reply, looking through the code some more you are right that this is still a reactive solution. It seems like `addCon` is called after the connections are initiated so it would essentially be a more extreme version of the connection manager. 

For outgoing connections, it is simple enough to put this check right before we dial a peer https://github.com/libp2p/go-libp2p-swarm/blob/master/swarm_dial.go#L293 . This would prevent any new outgoing connection from being created entirely. However for inbound connections this is a bit more complicated as there is no way to prevent an external peer from dialing us, from what I understand of the code the best place we can probably put a limit check is here https://github.com/libp2p/go-libp2p-swarm/blob/master/swarm_listen.go#L79. 

My main purpose for this PR was to prevent any connection notification handlers from being called if we exceed the peer limit. https://github.com/libp2p/go-libp2p-swarm/blob/master/swarm.go#L235 . Since this lead to a lot of the peer churn issues we were seeing when using the connection manager previously. 

I will look into implementing a much more strict connection manager for this, since it looks like adding this PR in might bring about some security risks to the protocol.

**Comment by nisdas** [comment]
Closing this PR as we have chosen to take an alternative approach of using a strict connection manager over here https://github.com/prysmaticlabs/prysm/pull/4110 .
